contracts/sg-ics721/src/execute.rs

```diff
@@ -36,45 +36,54 @@ impl Ics721Execute for SgIcs721Contract {
-                let class_data_result: StdResult<CollectionData> = from_binary(&data);
-                if class_data_result.is_err() {
-                    // this happens only for unknown class data, like source chain uses nft-transfer module
-                    env.contract.address.to_string()
-                } else {
-                    let class_data = class_data_result?;
-
-                    match class_data.owner {
-                        Some(owner) => convert_owner_chain_address(env, owner.as_str())?,
-                        None => env.contract.address.to_string(),
-                    }
-                }
-            }
-        };
-        to_binary(&sg721::InstantiateMsg {
+        // ics721 creator is used, in case no source owner in class data is provided (e.g. due to nft-transfer module).
+        let ics721_contract_info = deps
+            .querier
+            .query_wasm_contract_info(env.contract.address.to_string())?;
+        let mut instantiate_msg = sg721::InstantiateMsg {
             // Name of the collection MUST be class_id as this is how
             // we create a map entry on reply.
             name: class.id.clone().into(),
             symbol: class.id.clone().into(),
             minter: env.contract.address.to_string(),
             collection_info: sg721::CollectionInfo {
-                creator,
+                creator: ics721_contract_info.creator,
                 description: "".to_string(),
                 image: "https://arkprotocol.io".to_string(),
                 external_link: None,
                 explicit_content: None,
                 start_trading_time: None,
                 royalty_info: None,
             },
-        })
+        };
+
+        // unwrapped to collection data and in case of success, set creator, name and symbol
+        if let Some(binary) = class.data.clone() {
+            let class_data_result: StdResult<CollectionData> = from_json(binary);
+            if class_data_result.is_ok() {
+                let class_data = class_data_result?;
+                match class_data.owner {
+                    Some(owner) =>
+                    // owner from source chain is used
+                    {
+                        instantiate_msg.collection_info.creator =
+                            convert_owner_chain_address(env, owner.as_str())?
+                    }
+                    None =>
+                    // ics721 creator is used, in case of none
+                    {
+                        let ics721_contract_info = deps
+                            .querier
+                            .query_wasm_contract_info(env.contract.address.to_string())?;
+                        instantiate_msg.collection_info.creator = ics721_contract_info.creator;
+                    }
+                }
+                // set name and symbol
+                instantiate_msg.symbol = class_data.symbol;
```

Comment:
nice looks like you can take name and symbol right from the class data object?

Reply:
Some background here: ics721 need to store `class id <> nft contract` map entries (e.g. for determining incoming NFTs are back- (=burn) or forward transfer (=mint).

Before usage of `instantiate2` we **had** to store class id in collection, since address is unknown yet. and on instantiation reply we have address but no class id, so had to store it as part of collection.
Which led to the side effect frontends not having source name and symbol.

By using `instantiate2` ics721 now knows address - even before instantiation. So it can store name and symbol provided by source chain.